sklearn/utils/extmath.py

```diff
@@ -704,21 +706,37 @@ def _batch_mean_variance_update(X, old_mean, old_variance, old_sample_count):
 
     References
     ----------
-    T. Chan, G. Golub, R. LeVeque. Algorithms for computing the sample variance:
-        recommendations, The American Statistician, Vol. 37, No. 3, pp. 242-247
+    T. Chan, G. Golub, R. LeVeque. Algorithms for computing the sample
+        variance: recommendations, The American Statistician, Vol. 37, No. 3,
+        pp. 242-247
 
+    Also, see the sparse implementation of this in
+    `utils.sparsefuncs.incr_mean_variance_axis` and
+    `utils.sparsefuncs_fast.incr_mean_variance_axis0`
     """
+    # old = stats until now
+    # new = the current increment
+    # updated = the aggregated stats
+    last_sum = last_mean * last_sample_count
+    last_unnormalized_variance = last_variance * last_sample_count
+
+    new_sample_count = X.shape[0]
     new_sum = X.sum(axis=0)
-    new_variance = X.var(axis=0) * X.shape[0]
-    old_sum = old_mean * old_sample_count
-    n_samples = X.shape[0]
-    updated_sample_count = old_sample_count + n_samples
-    partial_variance = old_sample_count / (n_samples * updated_sample_count) * (
-        n_samples / old_sample_count * old_sum - new_sum) ** 2
-    unnormalized_variance = old_variance * old_sample_count + new_variance + \
-        partial_variance
-    return ((old_sum + new_sum) / updated_sample_count,
-            unnormalized_variance / updated_sample_count,
+    new_unnormalized_variance = X.var(axis=0) * new_sample_count
+
+    updated_sample_count = last_sample_count + new_sample_count
+    if last_sample_count == 0:  # Avoid division by 0
```

Comment:
Just curious if this condition ever happens, because you have checked `hasattr(self, 'n_samples_seen_')` already.